apm-agent-plugins/apm-vertx/apm-vertx-3_6-plugin/src/main/java/co/elastic/apm/agent/vertx_3_6/wrapper/GenericHandlerWrapper.java

```diff
@@ -0,0 +1,52 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2021 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.vertx_3_6.wrapper;
+
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import io.vertx.core.Handler;
+
+public class GenericHandlerWrapper<T> implements Handler<T> {
+
+    private final Handler<T> actualHandler;
+    private final AbstractSpan parentSpan;
+
+    public GenericHandlerWrapper(AbstractSpan parentSpan, Handler<T> actualHandler) {
+        this.parentSpan = parentSpan;
+        this.actualHandler = actualHandler;
+    }
```

Comment:
To guard against situations where the parent has ended and is recycled before the `handle` method starts, we'll have to increment the reference count here and decrement it after activating the parent span.

There may be other places where this is relevant.

```suggestion
        parentSpan.incrementReferences();
    }
```